From 6473e9972626750d9d52ea32f2864a8e74b92dbf Mon Sep 17 00:00:00 2001
From: Angus CLARK <angus.clark@st.com>
Date: Wed, 22 Apr 2009 09:41:19 +0100
Subject: [PATCH] sh_stm: Disables 'pc-mode' for EMI NAND Banks

Refactors EMI pc-mode configuration from pata setup, and disables pc-mode when
configuring an EMI bank for NAND flash access.

Signed-off-by: Angus Clark <angus.clark@st.com>
---
 drivers/stm/emi.c |   50 +++++++++++++++++++++++++++++++-------------------
 1 files changed, 31 insertions(+), 19 deletions(-)

diff --git a/drivers/stm/emi.c b/drivers/stm/emi.c
index 41ebca9..de8e050 100644
--- a/drivers/stm/emi.c
+++ b/drivers/stm/emi.c
@@ -65,6 +65,33 @@ void emi_bank_configure(int bank, unsigned long data[4])
 		writel(data[i], emi_control + BANK_EMICONFIGDATA(bank, i));
 }
 
+void emi_config_pcmode(int bank, int pc_mode)
+{
+	int mask;
+
+	BUG_ON(!emi_initialised);
+
+	switch (bank) {
+	case 2:	/* Bank C */
+		mask = 1<<3;
+		break;
+	case 3:	/* Bank D */
+		mask = 1<<4;
+		break;
+	default:
+		mask = 0;
+		break;
+	}
+
+	if (mask) {
+		u32 val = readl(emi_control + EMI_GEN_CFG);
+		if (pc_mode)
+			val |= mask;
+		else
+			val &= (~mask);
+		writel(val, emi_control + EMI_GEN_CFG);
+	}
+}
 
 
 /*
@@ -147,26 +174,8 @@ void __init emi_config_pata(int bank, int pc_mode)
 	set_pata_read_timings(bank, 120, 35, 30, 20);
 	set_pata_write_timings(bank, 120, 35, 30);
 
-	switch (bank) {
-	case 2:	/* Bank C */
-		mask = 1<<3;
-		break;
-	case 3:	/* Bank D */
-		mask = 1<<4;
-		break;
-	default:
-		mask = 0;
-		break;
-	}
+	emi_config_pcmode(bank, pc_mode);
 
-	if (mask) {
-		u32 val = readl(emi_control + EMI_GEN_CFG);
-		if (pc_mode)
-			val |= mask;
-		else
-			val &= (~mask);
-		writel(val, emi_control + EMI_GEN_CFG);
-	}
 }
 
 static void __init set_nand_read_timings(int bank, int cycle_time,
@@ -214,6 +223,9 @@ void __init emi_config_nand(int bank, struct emi_timing_data *timing_data)
 			timing_data->wr_cycle_time,
 			timing_data->wr_oee_start,
 			timing_data->wr_oee_end);
+
+	/* Disable PC mode */
+	emi_config_pcmode(bank, 0);
 }
 
 #ifdef CONFIG_PM
-- 
1.6.0.6

